ci/run/bazel_sharding/bazel_sharding.py

```diff
@@ -0,0 +1,367 @@
+    # 2. Allocate the next biggest item into the first shard that is below the optimum
+    # 3. If there's no shard below optimium, choose the shard closest to optimum
+
+    shards: List[List[BazelRule]] = [list() for _ in range(count)]
+    optimum = (
+        sum(timeout * len(rules) for timeout, rules in rules_grouped_by_time) / count
+    )
+
+    def get_next_biggest_item() -> BazelRule:
+        item = None
+        for timeout, items in rules_grouped_by_time:
+            if items:
+                item = items.pop()
+                break
+        return item
+
+    def get_shard_index_to_add_to(item_to_add: BazelRule) -> int:
+        shard_indices_below_optimum = []
+        shard_index_right_above_optimum = None
+        shard_index_right_above_optimum_time = None
+        for i, shard in enumerate(shards):
+            shard_time = sum(rule.actual_timeout_s for rule in shard)
+            shard_time_with_item = shard_time + item_to_add.actual_timeout_s
+            if shard_time_with_item < optimum:
+                shard_indices_below_optimum.append(i)
+            elif (
+                shard_index_right_above_optimum is None
+                or shard_index_right_above_optimum_time > shard_time_with_item
+            ):
+                shard_index_right_above_optimum = i
+                shard_index_right_above_optimum_time = shard_time_with_item
+        if shard_indices_below_optimum:
+            return shard_indices_below_optimum[0]
+        return shard_index_right_above_optimum
+
+    item_to_add = get_next_biggest_item()
+    while item_to_add:
+        shard_index_to_add_to = get_shard_index_to_add_to(item_to_add)
+        shards[shard_index_to_add_to].append(item_to_add)
+        item_to_add = get_next_biggest_item()
+
+    # Sanity checks.
+    num_all_targets = sum(len(shard) for shard in shards)
+    assert (
+        num_all_targets == expected_num_all_targets
+    ), f"got {num_all_targets} targets, expected {expected_num_all_targets}"
+    all_targets_set = set()
+    for shard in shards:
+        all_targets_set = all_targets_set.union(set(shard))
+    assert len(all_targets_set) == num_all_targets, (
+        f"num of unique targets {len(all_targets_set)} "
+        f"doesn't match num of targets {num_all_targets}"
+    )
+    assert all_targets_set == set(all_targets_set), (
+        f"unique targets after sharding {len(all_targets_set)} "
+        f"doesn't match unique targets after sharding {num_all_targets}"
+    )
+
+    print(
+        f"get_targets_for_shard statistics:\n\tOptimum: {optimum} seconds\n"
+        + "\n".join(
+            (
+                f"\tShard {i}: {len(shard)} targets, "
+                f"{sum(rule.actual_timeout_s for rule in shard)} seconds"
+            )
+            for i, shard in enumerate(shards)
+        ),
+        file=sys.stderr,
+    )
+    return sorted([rule.name for rule in shards[index]])
+
+
+def main(
+    targets: List[str],
+    index: int,
```

Comment:
```suggestion
    targets: List[str],
    *,
    index: int,
```

Can we force kwargs here so we end up with nicer calls?